File: public/thing.js
const Gravity = 0.01

const AnimationRate = 16

const AnimationNotDone = 0
const AnimationAlmostDone = 1
const AnimationDone = 2

const AnimationFront = 0
const AnimationFrontSide = 1
const AnimationSide = 2
const AnimationBackSide = 3
const AnimationBack = 4

const DirectionNorth = 0
const DirectionNorthEast = 1
const DirectionEast = 2
const DirectionSouthEast = 3
const DirectionSouth = 4
const DirectionSouthWest = 5
const DirectionWest = 6
const DirectionNorthWest = 7
const DirectionCount = 8
const DirectionNone = 8

const DirectionToAngle = [
    0.0 * DegToRad,
    45.0 * DegToRad,
    90.0 * DegToRad,
    135.0 * DegToRad,
    180.0 * DegToRad,
    225.0 * DegToRad,
    270.0 * DegToRad,
    315.0 * DegToRad
]

class Thing {
    constructor() {
        this.World = null
        this.UID = ""
        this.SID = ""
        this.NID = ""
        this.Animation = null
        this.AnimationMod = 0
        this.AnimationFrame = 0
        this.X = 0
        this.Y = 0
        this.Z = 0
        this.Angle = 0
        this.DX = 0
        this.DY = 0
        this.DZ = 0
        this.OX = 0
        this.OY = 0
        this.OZ = 0
        this.MinBX = 0
        this.MinBY = 0
        this.MinBZ = 0
        this.MaxBX = 0
        this.MaxBY = 0
        this.MaxBZ = 0
        this.Ground = false
        this.Radius = 0
        this.Height = 0
        this.Speed = 0
        this.Health = 0
    }
    static LoadNewThing(world, uid, nid, x, y, z) {
        switch (uid) {
            case "you":
                return new You(world, nid, x, y, z)
            case "baron":
                return new Baron(world, nid, x, y, z)
            case "tree":
                return new Tree(world, nid, x, y, z)
        }
    }
    Save(x, y, z) {
        return "{u:" + this.UID + ",x:" + (this.X - x) + ",y:" + (this.Y - y) + ",z:" + (this.Z - z) + "}"
    }
    BlockBorders() {
        this.MinBX = Math.floor((this.X - this.Radius) * InverseBlockSize)
        this.MinBY = Math.floor(this.Y * InverseBlockSize)
        this.MinBZ = Math.floor((this.Z - this.Radius) * InverseBlockSize)
        this.MaxBX = Math.floor((this.X + this.Radius) * InverseBlockSize)
        this.MaxBY = Math.floor((this.Y + this.Height) * InverseBlockSize)
        this.MaxBZ = Math.floor((this.Z + this.Radius) * InverseBlockSize)
    }
    AddToBlocks() {
        for (let gx = this.MinBX; gx <= this.MaxBX; gx++) {
            for (let gy = this.MinBY; gy <= this.MaxBY; gy++) {
                for (let gz = this.MinBZ; gz <= this.MaxBZ; gz++) {
                    this.World.GetBlock(gx, gy, gz).AddThing(this)
                }
            }
        }
    }
    RemoveFromBlocks() {
        for (let gx = this.MinBX; gx <= this.MaxBX; gx++) {
            for (let gy = this.MinBY; gy <= this.MaxBY; gy++) {
                for (let gz = this.MinBZ; gz <= this.MaxBZ; gz++) {
                    this.World.GetBlock(gx, gy, gz).RemoveThing(this)
                }
            }
        }
    }
    UpdateAnimation() {
        this.AnimationMod++
        if (this.AnimationMod === AnimationRate) {
            this.AnimationMod = 0
            this.AnimationFrame++
            let len = this.Animation.length
            if (this.AnimationFrame === len - 1)
                return AnimationAlmostDone
            else if (this.AnimationFrame === len)
                return AnimationDone
        }
        return AnimationNotDone
    }
    NetUpdateState(_) {}
    TerrainCollisionY(world) {
        if (this.DY < 0) {
            let gx = Math.floor(this.X)
            let gy = Math.floor(this.Y)
            let gz = Math.floor(this.Z)
            let bx = Math.floor(gx * InverseBlockSize)
            let by = Math.floor(gy * InverseBlockSize)
            let bz = Math.floor(gz * InverseBlockSize)
            let tx = gx - bx * BlockSize
            let ty = gy - by * BlockSize
            let tz = gz - bz * BlockSize

            let tile = world.GetTileType(bx, by, bz, tx, ty, tz)
            if (TileClosed[tile]) {
                this.Y = gy + 1
                this.Ground = true
                this.DY = 0
            }
        }
    }
    Resolve(b) {
        let square = this.Radius + b.Radius
        if (Math.abs(this.X - b.X) > square || Math.abs(this.Z - b.Z) > square)
            return
        if (Math.abs(this.OX - b.X) > Math.abs(this.OZ - b.Z)) {
            if (this.OX - b.X < 0) this.X = b.X - square
            else this.X = b.X + square
            this.DX = 0.0
        } else {
            if (this.OZ - b.Z < 0) this.Z = b.Z - square
            else this.Z = b.Z + square
            this.DZ = 0.0
        }
    }
    Overlap(b) {
        let square = this.Radius + b.Radius
        return Math.abs(this.X - b.X) <= square && Math.abs(this.Z - b.Z) <= square
    }
    Damage(_) {}
    Integrate() {
        // OX and snapshot need to be different things
        // this.OX = this.X
        // this.OY = this.Y
        // this.OZ = this.Z

        // if (this.DX != 0.0 || this.DZ != 0.0) {
        //     this.X += this.DX
        //     this.Z += this.DZ

        //     let collided = []
        //     let searched = new Set()

        //     this.RemoveFromBlocks(world)

        //     for (let gx = this.MinBX; gx <= this.MaxBX; gx++) {
        //         for (let gy = this.MinBY; gy <= this.MaxBY; gy++) {
        //             for (let gz = this.MinBZ; gz <= this.MaxBZ; gz++) {
        //                 let block = world.GetBlock(gx, gy, gz)
        //                 for (let t = 0; t < block.thingCount; t++) {
        //                     let thing = block.things[t]
        //                     if (searched.has(thing)) continue
        //                     searched.add(thing)
        //                     if (this.Overlap(thing)) collided.push(thing)
        //                 }
        //             }
        //         }
        //     }

        //     while (collided.length > 0) {
        //         let closest = null
        //         let manhattan = Number.MAX_VALUE
        //         for (let i = 0; i < collided.length; i++) {
        //             let thing = collided[i]
        //             let dist = Math.abs(this.OX - thing.X) + Math.abs(this.OZ - thing.Z)
        //             if (dist < manhattan) {
        //                 manhattan = dist
        //                 closest = thing
        //             }
        //         }
        //         this.Resolve(closest)
        //         collided.splice(closest)
        //     }

        //     this.BlockBorders()
        //     this.AddToBlocks(world)

        //     this.DX = 0.0
        //     this.DZ = 0.0
        // }

        // if (!this.Ground || this.DY != 0.0) {
        //     this.DY -= GRAVITY
        //     this.Y += this.DY
        //     this.TerrainCollisionY(world)

        //     this.RemoveFromBlocks(world)
        //     this.BlockBorders()
        //     this.AddToBlocks(world)
        // }
    }
    Render(interpolation, spriteBuffer, camX, camZ, camAngle) {
        let vx = this.OX + interpolation * (this.X - this.OX)
        let vy = this.OY + interpolation * (this.Y - this.OY)
        let vz = this.OZ + interpolation * (this.Z - this.OZ)

        let sin = camX - vx
        let cos = camZ - vz
        let length = Math.sqrt(sin * sin + cos * cos)
        sin /= length
        cos /= length

        let angle = camAngle - this.Angle
        if (angle < 0) angle += Tau

        let direction
        let mirror

        const AngleA = 337.5 * DegToRad
        const AngleB = 292.5 * DegToRad
        const AngleC = 247.5 * DegToRad
        const AngleD = 202.5 * DegToRad
        const AngleE = 157.5 * DegToRad
        const AngleF = 112.5 * DegToRad
        const AngleG = 67.5 * DegToRad
        const AngleH = 22.5 * DegToRad

        if (angle > AngleA) {
            direction = AnimationFront
            mirror = false
        } else if (angle > AngleB) {
            direction = AnimationFrontSide
            mirror = false
        } else if (angle > AngleC) {
            direction = AnimationSide
            mirror = false
        } else if (angle > AngleD) {
            direction = AnimationBackSide
            mirror = false
        } else if (angle > AngleE) {
            direction = AnimationBack
            mirror = false
        } else if (angle > AngleF) {
            direction = AnimationBackSide
            mirror = true
        } else if (angle > AngleG) {
            direction = AnimationSide
            mirror = true
        } else if (angle > AngleH) {
            direction = AnimationFrontSide
            mirror = true
        } else {
            direction = AnimationFront
            mirror = false
        }

        let sprite = this.Animation[this.AnimationFrame][direction]

        if (mirror)
            Render3.MirrorSprite(spriteBuffer[this.SID], vx, vy, vz, sin, cos, sprite)
        else
            Render3.Sprite(spriteBuffer[this.SID], vx, vy, vz, sin, cos, sprite)
    }
}

File: thing.go
package main

import (
	"bytes"
	"math"
	"strings"
)

// Thing constants
const (
	AnimationRate       = 5
	Gravity             = 0.01
	AnimationNotDone    = 0
	AnimationAlmostDone = 1
	AnimationDone       = 2
)

// UID constants
const (
	HumanUID  = uint16(0)
	BaronUID  = uint16(1)
	TreeUID   = uint16(2)
	PlasmaUID = uint16(3)
)

// Thing variables
var (
	ThingNetworkNum = uint16(0)
)

// Thing struct
type Thing struct {
	World               *World
	UID                 uint16
	NID                 uint16
	Animation           int
	AnimationMod        int
	AnimationFrame      int
	X, Y, Z             float32
	Angle               float32
	DeltaX, DeltaY, DeltaZ          float32
	OldX, OldZ          float32
	MinBX, MinBY, MinBZ int
	MaxBX, MaxBY, MaxBZ int
	Ground              bool
	Radius              float32
	Height              float32
	Speed               float32
	Health              int
	DeltaMoveXZ         bool
	DeltaMoveY          bool
	Update              func() bool
	Damage              func(int)
	Save                func(data *strings.Builder)
	BinarySave          func(raw *bytes.Buffer)
	Snap                func(raw *bytes.Buffer) int
}

// NextNID func
func NextNID() uint16 {
	ThingNetworkNum++
	return ThingNetworkNum
}

// LoadNewThing func
func LoadNewThing(world *World, uid string, x, y, z float32) *Thing {
	switch uid {
	case "spawn":
		world.SpawnYouX = x
		world.SpawnYouY = y
		world.SpawnYouZ = z
	case "baron":
		return NewBaron(world, x, y, z).Thing
	case "tree":
		return NewTree(world, x, y, z)
	}
	return nil
}

// NopUpdate func
func (me *Thing) NopUpdate() bool {
	return false
}

// NopSnap func
func (me *Thing) NopSnap(raw *bytes.Buffer) int {
	return 0
}

// NopDamage func
func (me *Thing) NopDamage(amount int) {
}

// BlockBorders func
func (me *Thing) BlockBorders() {
	me.MinBX = int((me.X - me.Radius) * InverseBlockSize)
	me.MinBY = int(me.Y * InverseBlockSize)
	me.MinBZ = int((me.Z - me.Radius) * InverseBlockSize)
	me.MaxBX = int((me.X + me.Radius) * InverseBlockSize)
	me.MaxBY = int((me.Y + me.Height) * InverseBlockSize)
	me.MaxBZ = int((me.Z + me.Radius) * InverseBlockSize)
}

// AddToBlocks func
func (me *Thing) AddToBlocks() {
	for gx := me.MinBX; gx <= me.MaxBX; gx++ {
		for gy := me.MinBY; gy <= me.MaxBY; gy++ {
			for gz := me.MinBZ; gz <= me.MaxBZ; gz++ {
				me.World.GetBlock(gx, gy, gz).AddThing(me)
			}
		}
	}
}

// RemoveFromBlocks func
func (me *Thing) RemoveFromBlocks() {
	for gx := me.MinBX; gx <= me.MaxBX; gx++ {
		for gy := me.MinBY; gy <= me.MaxBY; gy++ {
			for gz := me.MinBZ; gz <= me.MaxBZ; gz++ {
				me.World.GetBlock(gx, gy, gz).RemoveThing(me)
			}
		}
	}
}

// UpdateAnimation func
func (me *Thing) UpdateAnimation() int {
	me.AnimationMod++
	if me.AnimationMod == AnimationRate {
		me.AnimationMod = 0
		me.AnimationFrame++
		if me.AnimationFrame == me.Animation-1 {
			return AnimationAlmostDone
		} else if me.AnimationFrame == me.Animation {
			return AnimationDone
		}
	}
	return AnimationNotDone
}

// TerrainCollisionXZ func
func (me *Thing) TerrainCollisionXZ() {
	minGX := int((me.X - me.Radius))
	minGY := int(me.Y)
	minGZ := int((me.Z - me.Radius))
	maxGX := int((me.X + me.Radius))
	maxGY := int((me.Y + me.Height))
	maxGZ := int((me.Z + me.Radius))

	minBX := int(float32(minGX) * InverseBlockSize)
	minBY := int(float32(minGY) * InverseBlockSize)
	minBZ := int(float32(minGZ) * InverseBlockSize)

	minTX := minGX - minBX*BlockSize
	minTY := minGY - minBY*BlockSize
	minTZ := minGZ - minBZ*BlockSize

	world := me.World

	bx := minBX
	tx := minTX
	for gx := minGX; gx <= maxGX; gx++ {
		by := minBY
		ty := minTY
		for gy := minGY; gy <= maxGY; gy++ {
			bz := minBZ
			tz := minTZ
			for gz := minGZ; gz <= maxGZ; gz++ {
				block := world.GetBlock(bx, by, bz)
				tile := block.GetTileTypeUnsafe(tx, ty, tz)
				if TileClosed[tile] {
					xx := float32(gx)
					closeX := me.X
					if closeX < xx {
						closeX = xx
					} else if closeX > xx+1 {
						closeX = xx + 1
					}

					zz := float32(gz)
					closeZ := me.Z
					if closeZ < zz {
						closeZ = zz
					} else if closeZ > zz+1 {
						closeZ = zz + 1
					}

					dxx := me.X - closeX
					dzz := me.Z - closeZ
					dist := dxx*dxx + dzz*dzz

					if dist > me.Radius*me.Radius {
						continue
					}
					dist = float32(math.Sqrt(float64(dist)))
					if dist == 0 {
						dist = 1
					}
					mult := me.Radius / dist
					me.X += dxx*mult - dxx
					me.Z += dzz*mult - dzz
				}
				tz++
				if tz == BlockSize {
					tz = 0
					bz++
				}
			}
			ty++
			if ty == BlockSize {
				ty = 0
				by++
			}
		}
		tx++
		if tx == BlockSize {
			tx = 0
			bx++
		}
	}
}

// TerrainCollisionY func
func (me *Thing) TerrainCollisionY() {
	if me.DeltaY < 0 {
		gx := int(me.X)
		gy := int(me.Y)
		gz := int(me.Z)
		bx := int(me.X * InverseBlockSize)
		by := int(me.Y * InverseBlockSize)
		bz := int(me.Z * InverseBlockSize)
		tx := gx - bx*BlockSize
		ty := gy - by*BlockSize
		tz := gz - bz*BlockSize

		tile := me.World.GetTileType(bx, by, bz, tx, ty, tz)
		if TileClosed[tile] {
			me.Y = float32(gy + 1)
			me.Ground = true
			me.DeltaY = 0
		}
	}
}

// Resolve func
func (me *Thing) Resolve(b *Thing) {
	square := me.Radius + b.Radius
	absx := Abs(me.X - b.X)
	absz := Abs(me.Z - b.Z)
	if absx > square || absz > square {
		return
	}
	x := me.OldX - b.X
	z := me.OldZ - b.Z
	if Abs(x) > Abs(z) {
		if x < 0 {
			me.X = b.X - square
		} else {
			me.X = b.X + square
		}
		me.DeltaX = 0.0
	} else {
		if z < 0 {
			me.Z = b.Z - square
		} else {
			me.Z = b.Z + square
		}
		me.DeltaZ = 0.0
	}
}

// Overlap func
func (me *Thing) Overlap(b *Thing) bool {
	square := me.Radius + b.Radius
	return Abs(me.X-b.X) <= square && Abs(me.Z-b.Z) <= square
}

// ApproximateDistance func
func (me *Thing) ApproximateDistance(other *Thing) float32 {
	dx := Abs(me.X - other.X)
	dy := Abs(me.Z - other.Z)
	if dx > dy {
		return dx + dy - dy*0.5
	}
	return dx + dy - dx*0.5
}

// IntegrateXZ func
func (me *Thing) IntegrateXZ() {
	me.OldX = me.X
	me.OldZ = me.Z

	me.X += me.DeltaX
	me.Z += me.DeltaZ
	me.DeltaMoveXZ = true

	collided := make([]*Thing, 0)
	searched := make(map[*Thing]bool)

	me.RemoveFromBlocks()
	me.BlockBorders()

	for gx := me.MinBX; gx <= me.MaxBX; gx++ {
		for gy := me.MinBY; gy <= me.MaxBY; gy++ {
			for gz := me.MinBZ; gz <= me.MaxBZ; gz++ {
				block := me.World.GetBlock(gx, gy, gz)
				for t := 0; t < block.ThingCount; t++ {
					thing := block.Things[t]
					if _, ok := searched[thing]; !ok {
						searched[thing] = true
						if me.Overlap(thing) {
							collided = append(collided, thing)
						}
					}
				}
			}
		}
	}

	for len(collided) > 0 {
		closest := 0
		manhattan := float32(math.MaxFloat32)
		for i := 0; i < len(collided); i++ {
			thing := collided[i]
			dist := Abs(me.OldX-thing.X) + Abs(me.OldZ-thing.Z)
			if dist < manhattan {
				manhattan = dist
				closest = i
			}
		}
		me.Resolve(collided[closest])
		copy(collided[closest:], collided[closest+1:])
		collided[len(collided)-1] = nil
		collided = collided[:len(collided)-1]
	}

	me.TerrainCollisionXZ()

	me.BlockBorders()
	me.AddToBlocks()

	me.DeltaX = 0.0
	me.DeltaZ = 0.0
}

// IntegrateY func
func (me *Thing) IntegrateY() {
	if !me.Ground {
		me.DeltaY -= Gravity
		me.Y += me.DeltaY
		me.DeltaMoveY = true
		me.TerrainCollisionY()

		me.RemoveFromBlocks()
		me.BlockBorders()
		me.AddToBlocks()
	}
}
